### DISCUSSION TITLE
Make Craft CMS 4 faster

### DISCUSSION THREAD

**Discussion by vitalijalbu** [craftcms/cms#12118] [General]
Hi guys, l'es discuss. On every craft 4+ version I found speed leaks 

Here is craft 3.7 with 1200+ entries (blog entries, which have lots of categories, tags, etc...)
<img width="1447" alt="Schermata 2022-10-13 alle 17 34 23" src="https://user-images.githubusercontent.com/40239902/195641537-695469e3-dd8a-4b79-b66a-cf08faca6d63.png">

Here is asimple craft 4.2 with 29 entries (simple pages and service lists  etc...)
<img width="1447" alt="Schermata 2022-10-13 alle 17 34 34" src="https://user-images.githubusercontent.com/40239902/195641821-d55d98fd-799b-4d72-8ff7-fb8fa4bbb810.png">

As you can see craft cms 4 response time is 3x times higher than 3

**Comment by vitalijalbu** [comment]
Sidebar slideout to edit entry also is slow, user must wait 2-3 secs to see output data. Not so acceptable for non-dev users

https://user-images.githubusercontent.com/40239902/195642779-58b77ca8-100b-49e6-a513-b289c255f957.mov

**Comment by brandonkelly** [comment]
In general Craft 4 is faster than Craft 3. One case where it might not be is when Dev Mode is enabled, though. Is that the case here? If so try disabling it.

**Comment by RyanRoberts** [comment]
I've noticed a significant slowdown in TTFB with sites updated to Craft 4 as well. Here's one very simple site before: https://www.webpagetest.org/result/221103_AiDcQC_BPV/ and after: https://www.webpagetest.org/result/221103_AiDcNV_C8S/

Other Craft 4 sites I have on a completely different server demonstrate the same TTFB slowdown, even after stripping a page right back while testing. Curiously this wasn't the case with v4 sites initially, here's four very different sites that suddenly slowed https://media.yep.digital/kxFFwV

I've tried everything from upgrading the server to building a new project from scratch and deploying to the same server. All my Craft 4 sites just seem to be slower.

I'll get in touch with support with a copy of a site, logs, db dump, etc

**Reply by brandonkelly** [reply]
Would you be able to share the Craft 3 and 4 database backups for a site you’re seeing this with, so we can investigate? If so, please send them to support@craftcms.com.

**Reply by RyanRoberts** [reply]
Will do, I have a couple more to upgrade so I'll keep the Craft 3 backups.

**Comment by thupsi** [comment]
For a while now I was also under the impression that the entry index in the cp of Craft 4 was a little sluggish. I hadn't done any measurements but I was seeing the loading "spinner" for longer than I would like. I'm not sure if this was the case from the initial 4.0 version or something changed in later versions.

Anyhow, I just upgraded a Craft 3 site and went on to test the 3 and 4 versions side by side. Just the entry index, since this is where I'm experiencing a slower than desired performance. The load speed as measured was just a little worse in Craft 4, but what is of more concern is **the time it takes to visually populate the index** (so this is on top of the initial page load). There is a noticeable difference in favour of Craft 3 without needing to see the measured page speed.

Here is Craft 3.7.59:


https://user-images.githubusercontent.com/19857599/200312182-a34dc9d2-f56a-493b-b249-88d47d1efce4.mp4


And here is Craft 4.3.1




https://user-images.githubusercontent.com/19857599/200312494-32b05581-68a0-47a7-ae9b-ce3afcc174c7.mp4



Dev Mode was off in both cases. As you can see, I just kept refreshing the same index on both sites. In Craft 3 the index populates almost instantly. In Craft 4 you can see the spinner icon for a second or so. It's not the end of the world, but it is noticeable and prevents the cp from feeling snappy.

Keep in mind that my own observations are exclusively about the entry index. Haven't measured, but also haven't noticed, a perfomance degradation in any other part of the cp, or in the front end.

**Reply by brandonkelly** [reply]
Would you be able to share the Craft 3 and 4 database backups for a site you’re seeing this with, so we can investigate? If so, please send them to support@craftcms.com.

**Reply by wsydney76** [reply]
Regarding element indexes in Craft 4, we are seeing a performance hit when relation fields are included.

For example in our live settings with two relation fields (one image, one entries field) we see for `get-elements` ,100 entries:

240 queries, 571 msec total (which is still quite ok, so didn't look into it until now.)

Adding the author field to the mix, we have:

642 queries, 1088 msec total

Removing all relation fields there are:

45 (!) queries, 253 msec.

So may be something is wrong with eager loading stuff?

/Aylin

**Reply by thupsi** [reply]
I just sent the DB backups to support.

Also, yep, **removing the author field as suggested by @wsydney76 made a noticeable difference**.

**Comment by brandonkelly** [comment]
Craft 4.3.4 reverted a change that turned out to be the cause of a lot of performance problems in Craft 4. (991396e42269463790a972dc86b9cf7bbb71caca)

**Reply by thupsi** [reply]
Hi! I've been in touch regarding this discussion via email (thanks for all the prompt responses from the team).

While I want to again state that this isn't a serious issue for me, I felt it's worth investigating further, so I finally found time to do more testing. 

My goal was to measure and quantify the perfomance diff in the entries index between a Craft 4 and Craft 3 install. 

For context:
- I have the same project on both versions.
- I have disabled `dev mode`.
- I'm using Chrome's "Performance insights" panel as a measuring tool.
- I'm specifically looking at **LCP** (Largest Contentfull Paint) and **TTI** (Time to Interactive) as my most important metrics.
- I repeated the measurements multiple times to ensure any perfomance diffs were consistent.

**Scenario 1 (No columns enabled, only entry title visible)**

| Version| LCP| TTI|
|--------|--------|--------|
| Craft 3| 0.51s| 0.36s|
| Craft 4| 0.74s| 0.72s| )

**Scenario 2 (Author column enabled)**

| Version| LCP| TTI|
|--------|--------|--------|
| Craft 3| 0.6s| 1.96s|
| Craft 4| 1.04s| 1.97s| )

**Scenario 3 (Post date, Link and Author columns enabled)**

| Version| LCP| TTI|
|--------|--------|--------|
| Craft 3| 0.7s| 1.83s|
| Craft 4| 1.15s| 2.04s| )

There seems indeed to be a small difference in performance between the two versions when it comes to displaying entry indexes. My subjective experience when navigating the CP, without doing any measurements, also is consistent with this conclusion. The way I can best describe it is that the index seems to be visually populated with entries slightly faster in Craft 3. That probably corelates to LCP.

Of course this is only part of the CP experience. For example saving complex entries/drafts is MUCH faster in Craft 4.

Anyway, I don't know if this is of any real importance (the perf difference is not that large, but it is noticeable for me), and if it's the price to pay for having filters and such in the index, then no complaints there! 

Just wanted to chime in with my observations. Thanks for reading 🙂

**Reply by vitalijalbu** [reply]
Yes, everybody knows that craft3 was more faster and performant... let's see if that will change in craft 5 :D see my video: https://github.com/craftcms/cms/discussions/12118#discussioncomment-3871475

**Reply by thupsi** [reply]
I wouldn't say that Craft 3 is faster in general, it's the oposite in some cases, like when saving entries or creating drafts with complex nested fields, where Craft 4 has applied some very nice optimizations. 

That being said, I think it's important that we report when and where we notice any performance degradation, because it is surelly impossible for the team to test all possible configurations and scenarios.

**Reply by brandonkelly** [reply]
Another thing that can drastically impact performance is whether Composer’s generated autoloader is optimized. The optimization process can be very resource intensive, so we opt out of it for Control Panel-triggered Composer operations (e.g. updating Craft/plugins).

To ensure your install is optimized, run `composer dump-autoload -o`.

@vitalijalbu That video predates 4.3.4.

**Reply by thupsi** [reply]
> Another thing that can drastically impact performance is whether Composer’s generated autoloader is optimized. The optimization process can be very resource intensive, so we opt out of it for Control Panel-triggered Composer operations (e.g. updating Craft/plugins).
> 
> To ensure your install is optimized, run `composer dump-autoload -o`.

That did not make a difference in my case. Guess my Composer autoloader was allready optimized.
